rclpy/rclpy/qos_event.py

```diff
@@ -0,0 +1,172 @@
+
+Mirrors rmw_liveliness_lost_status_t from rmw/types.h
+"""
+QoSLivelinessLostInfo = namedtuple(
+    'QoSLivelinessLostInfo',
+    ('total_count', 'total_count_change'))
+
+
+class SubscriptionEventCallbacks:
+    """Container to provide middleware event callbacks for a Subscription."""
+
+    def __init__(
+        self,
+        *,
+        deadline: Callable[[QoSRequestedDeadlineMissedInfo], None] = None,
+        liveliness: Callable[[QoSLivelinessChangedInfo], None] = None,
+    ) -> None:
+        """
+        Constructor.
+
+        :param deadline: A user-defined callback that is called when a topic misses our
+            requested Deadline.
+        :param liveliness: A user-defined callback that is called when the Liveliness of
+            a Publisher on subscribed topic changes.
+        """
+        self.deadline = deadline
+        self.liveliness = liveliness
+
+
+class PublisherEventCallbacks:
+    """Container to provide middleware event callbacks for a Publisher."""
+
+    def __init__(
+        self,
+        *,
+        deadline: Callable[[QoSOfferedDeadlineMissedInfo], None] = None,
+        liveliness: Callable[[QoSLivelinessLostInfo], None] = None
+    ) -> None:
+        """
+        Constructor.
+
+        :param deadline: A user-defined callback that is called when the Publisher misses
+            its offered Deadline.
+        :param liveliness: A user-defined callback that is called when this Publisher
+            fails to signal its Liveliness and is reported as not-alive.
+        """
+        self.deadline = deadline
+        self.liveliness = liveliness
+
+
+class QoSEventHandler(Waitable):
+    """Waitable type to handle QoS events."""
+
+    def __init__(
+        self,
+        *,
+        callback_group,
+        callback,
+        event_type,
+        parent_handle,
+    ):
+        # Waitable init adds self to callback_group
+        super().__init__(callback_group)
+        self.event_type = event_type
+        self.callback = callback
+
+        self._parent_handle = parent_handle
+        self._event_handle = _rclpy.rclpy_create_event(event_type, parent_handle)
+        self._event_is_ready = False
+        self._event_index = None
+
+    # Start Waitable API
+    def is_ready(self, wait_set):
+        """Return True if entities are ready in the wait set."""
+        if _rclpy.rclpy_wait_set_is_ready('event', wait_set, self._event_index):
+            self._event_is_ready = True
+        return self._event_is_ready
+
+    def take_data(self):
+        """Take stuff from lower level so the wait set doesn't immediately wake again."""
+        if self._event_is_ready:
+            self._event_is_ready = False
+            return _rclpy.rclpy_take_event(
+                self._event_handle, self._parent_handle, self.event_type)
+        return None
+
+    async def execute(self, taken_data):
+        """Execute work after data has been taken from a ready wait set."""
+        self.callback(taken_data)
```

Comment:
Based on usage (`executors.py`) it doesn't happen because the calling order is always `is_ready True -> take_data -> execute` - but I'll add the check for thoroughness, definitely doesn't hurt